### PR TITLE
Fixed dangerous cast. 

### PR DESCRIPTION
**Pull request by lancerchao**
Taken from OpenSSL

### CODE REVIEW

File: af_ktls.c [hunk 1]
```diff
@@ -286,12 +286,15 @@ static inline struct tls_sock *tls_sk(struct sock *sk)
...
-	seq_h++;
-	*seqno = cpu_to_be64(seq_h);
+    int i;
+
+    for (i = 7; i >= 0; i--) {
+        ++seq[i];
+        if (seq[i] != 0)
+            break;
+    }
 }
```
Comment (nmav) [inline af_ktls.c]: That will take some more instructions but looks fine as it is not really a bottleneck. What I noticed though is that this function does not check for wrap-arounds, and it increments the same for TLS and DTLS. In DTLS only the rightmost 48-bits are incremented, the rest is the epoch.

Comment (lancerchao) [inline af_ktls.c]: > this function does not check for wrap-arounds, and it increments the same for TLS and DTLS. In DTLS only the rightmost 48-bits are incremented, the rest is the epoch.

Good catch! I added a fix for that, although I'm not 100% sure that it is necessary. I think that userspace should make sure that the sequence numbers doesn't wrap.

File: af_ktls.c [hunk 2]
```diff
@@ -286,15 +286,21 @@ static inline struct tls_sock *tls_sk(struct sock *sk)
...
-    }
+	for (i = 7; i >= 0; i--) {
+		++seq[i];
+		if (seq[i] != 0)
+			break;
+	}
+	/*
+	 * Check for overflow. If overflowed, connection must either rehandshake
+	 * or disconnect. Raise an error and notify userspace
+	 */
```
Comment (nmav) [inline af_ktls.c]: A nitpick on the comment. If the overflow occurs you can no longer rehandshake as you have no more numbers to use to exchange the handshake packets. At this point you can only terminate the session.

**Comment by nmav** [comment]
LGTM

File: af_ktls.c (hunk 2 above)
Comment (lancerchao) [inline af_ktls.c]: > A nitpick on the comment. If the overflow occurs you can no longer rehandshake as you have no more numbers to use to exchange the handshake packets. At this point you can only terminate the session.

Yes! Thanks for the correction.

**Comment by fridex** [comment]
Looks good to me as well. Cannot merge now because of ```tls_err_abort()``` which is introduced in #62. Will do the review.

**Comment by djwatson** [comment]
LGTM.  Let's merge this and just rebase #62 on top?

**Comment by fridex** [comment]
The static keyword was removed. I would merge it without overflow check and leave it for the next patch?

I'm not sure about the overflow handling like implemented here from user space POV. The user space has to trigger rehandshake. There needs to be checked how GnuTLS and OpenSSL behaves in this situation.

IIRC, GnuTLS does handshake explicitly whereas OpenSSL implicitly/transparently during communication. I'm not sure if we trigger rehandshake just by propagating an error this way.

**Comment by djwatson** [comment]
Ah I see.  Yea I would merge without overflow and open a separate issue for overflow.

**Comment by nmav** [comment]
> I'm not sure about the overflow handling like implemented here from user space POV. The user space has to trigger rehandshake. There needs to be checked how GnuTLS and OpenSSL behaves in this situation.

The overflow check must be present as a last defense to protect userspace. If userspace continues sending packets after an overflow occurred that will allow the decryption of the transmitted data, due to the way CTR mode (used by GCM) operates. Both openssl and gnutls do overflow checks, so applications which use them and switch to af_ktls would expect the same.

**Pull request merged**